frame/staking/src/offchain_election.rs

```diff
@@ -73,42 +88,112 @@ pub(crate) const DEFAULT_LONGEVITY: u64 = 25;
 /// Returns `Ok(())` if offchain worker should happen, `Err(reason)` otherwise.
 pub(crate) fn set_check_offchain_execution_status<T: Config>(
 	now: T::BlockNumber,
-) -> Result<(), &'static str> {
+	threshold: T::BlockNumber,
+) -> Result<(), OffchainElectionError> {
 	let storage = StorageValueRef::persistent(&OFFCHAIN_HEAD_DB);
-	let threshold = T::BlockNumber::from(OFFCHAIN_REPEAT);
-
-	let mutate_stat =
-		storage.mutate::<_, &'static str, _>(|maybe_head: Option<Option<T::BlockNumber>>| {
-			match maybe_head {
-				Some(Some(head)) if now < head => Err("fork."),
-				Some(Some(head)) if now >= head && now <= head + threshold => {
-					Err("recently executed.")
-				}
-				Some(Some(head)) if now > head + threshold => {
-					// we can run again now. Write the new head.
-					Ok(now)
-				}
-				_ => {
-					// value doesn't exists. Probably this node just booted up. Write, and run
-					Ok(now)
-				}
+
+	let mutate_stat = storage.mutate(|maybe_head: Option<Option<T::BlockNumber>>| {
+		match maybe_head {
+			Some(Some(head)) if now < head => Err(OffchainElectionError::Fork),
+			Some(Some(head)) if now >= head && now <= head + threshold => {
+				Err(OffchainElectionError::TooRecent)
+			}
+			Some(Some(head)) if now > head + threshold => {
+				// we can allow again now. Write the new head.
+				Ok(now)
+			}
+			_ => {
+				// value doesn't exists. Probably this node just booted up. Write, and allow.
+				Ok(now)
 			}
-		});
+		}
+	});
 
+	crate::log!(trace, "attempting to acquire the OCW lock at {:?} = {:?}", now, mutate_stat);
 	match mutate_stat {
 		// all good
 		Ok(Ok(_)) => Ok(()),
 		// failed to write.
-		Ok(Err(_)) => Err("failed to write to offchain db."),
+		Ok(Err(_)) => Err(OffchainElectionError::DBWriteFailed),
 		// fork etc.
 		Err(why) => Err(why),
 	}
 }
 
+/// Storage path for the solution `call`.
+pub(crate) const OFFCHAIN_QUEUED_CALL: &[u8] = b"parity/staking-election/call";
+
+/// Save a given call OCW storage.
+pub(crate) fn save_solution<T: Config>(call: Call<T>) -> Result<(), OffchainElectionError> {
+	let storage = StorageValueRef::persistent(&OFFCHAIN_QUEUED_CALL);
+	let set_outcome = storage.mutate::<_, OffchainElectionError, _>(|maybe_call| {
+		match maybe_call {
```

Comment:
The `match` is still very much redundant, maybe let's just move the comment above the entire `mutate` call` and just call `mutate::<...>(|_| Ok(call))`?